api/src/routes/settings.ts

```diff
@@ -92,6 +119,107 @@ export const settingRoutes: FastifyPluginCallbackTypebox = (
     }
   );
 
+  fastify.put(
+    '/update-my-email',
+    {
+      schema: schemas.updateMyEmail,
+      // We need to customize the responses to validation failures:
+      attachValidation: true
+    },
+    async (req, reply) => {
+      if (req.validationError) {
+        void reply.code(400);
+        return { message: 'Email format is invalid', type: 'danger' } as const;
+      }
+
+      const user = await fastify.prisma.user.findUniqueOrThrow({
+        where: { id: req.session.user.id },
+        select: {
+          email: true,
+          emailVerifyTTL: true,
+          newEmail: true,
+          emailVerified: true,
+          emailAuthLinkTTL: true
+        }
+      });
+      const newEmail = req.body.email.toLowerCase();
+      const currentEmailFormatted = user.email.toLowerCase();
+      const isVerifiedEmail = user.emailVerified;
+      const isOwnEmail = newEmail === currentEmailFormatted;
+      if (isOwnEmail && isVerifiedEmail) {
+        void reply.code(400);
+        return {
+          type: 'info',
+          message: `${newEmail} is already associated with this account.
+You can update a new email address instead.`
+        } as const;
+      }
+
+      const isResendUpdateToSameEmail =
+        newEmail === user.newEmail?.toLowerCase();
+      const isLinkSentWithinLimitTTL = getWaitMessage(user.emailVerifyTTL);
+
+      if (isResendUpdateToSameEmail && isLinkSentWithinLimitTTL) {
+        void reply.code(429);
+        return {
+          type: 'info',
+          message: `We have already sent an email confirmation request to ${newEmail}.
+${isLinkSentWithinLimitTTL}`
+        } as const;
+      }
+
+      const isEmailAlreadyTaken =
+        (await fastify.prisma.user.count({ where: { email: newEmail } })) > 0;
+
+      if (isEmailAlreadyTaken && !isOwnEmail) {
+        void reply.code(400);
+        return {
+          type: 'info',
+          message: `${newEmail} is already associated with another account.`
+        } as const;
+      }
+
+      // ToDo(MVP): email the new email and wait user to confirm it, before we update the user schema.
+      try {
+        await fastify.prisma.user.update({
+          where: { id: req.session.user.id },
+          data: {
+            newEmail,
+            emailVerified: false,
+            emailVerifyTTL: new Date()
+          }
+        });
+
+        // TODO: combine emailVerifyTTL and emailAuthLinkTTL? I'm not sure why
+        // we need emailVeriftyTTL given that the main thing we want is to
+        // restrict the rate of attempts and the emailAuthLinkTTL already does
+        // that.
+        const tooManyRequestsMessage = getWaitMessage(user.emailAuthLinkTTL);
+
+        if (tooManyRequestsMessage) {
+          void reply.code(429);
+          return {
+            type: 'info',
+            message: tooManyRequestsMessage
+          } as const;
+        }
```

Comment:
From what I can tell in the old api, we run the `tooManyRequestsMessage` before updating the email. We should do the same here.

That is, move the `if (tooManyRequestsMessage)` to be before the `fastify.prisma.user.update`